pkg/sql/pg_catalog_test.go

```diff
@@ -0,0 +1,379 @@
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// This test compares a dump generated from PostgreSQL with the pg_catalog
+// and compares it with the current pg_catalog at cockroach db skipping
+// all the known diffs:
+//
+// cd pkg/sql
+// go test -run TestPGCatalog
+//
+// If you want to re-create the known (expected) diffs with the current differences
+// add -rewrite-diffs flag when running this test:
+//
+// cd pkg/sql
+// go test -run TestPGCatalog -rewrite-diffs
+//
+// To create the postgres dump file see pkg/cmd/generate-pg-catalog/main.go:
+//
+// cd pkg/cmd/generate-pg-catalog/
+// go run main.go > ../../sql/testdata/pg_catalog_tables
+package sql
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/errors/oserror"
+)
+
+// This constants enumerates indexes in pg_catalog_tables
+const (
+	testTableIndex = iota
+	testColumnIndex
+	testTypeIndex
+	testTypeOidIndex
+)
+
+// This query uses udt_name::regtype instead of data_type column because
+// data_type only says "ARRAY" but does not say which kind of array it is.
+const getPgCatalogSQL = `
+	SELECT table_name, column_name, udt_name::regtype data_type, udt_name::regtype::Oid data_type_oid
+	FROM information_schema.columns 
+	WHERE table_schema = 'pg_catalog'
+	ORDER BY 1, 2;
+`
+
+// Test data files
+const (
+	pgCatalogDump = "pg_catalog_tables"          // PostgreSQL pg_catalog schema
+	expectedDiffs = "pg_catalog_test-diffs.json" // Contains expected difference between postgres and cockroach
+	testdata      = "testdata"                   // testdata directory
+)
+
+// When running test with -rewrite-diffs test will pass and re-create pg_catalog_test-diffs.json
+var rewriteFlag = flag.Bool("rewrite-diffs", false, "This will re-create the expected diffs file")
+
+// summary will keep accountability for any unexpected difference and report it in the log
+type summary struct {
+	missingTables        int
+	missingColumns       int
+	mismatchDatatypesOid int
+}
+
+// TestColumn is a structure which contains a small description about the datatype of a column, but this can also be
+// used as a diff information if populating ExpOid and ExpDataType. Fields are exported for Marshaling purposes.
+type TestColumn struct {
+	Oid         uint32  `json:"oid"`
+	DataType    string  `json:"dataType"`
+	ExpOid      *uint32 `json:"expOid"`
+	ExpDataType *string `json:"expDataType"`
+}
+
+// testColumns maps column names with datatype description
+type testColumns map[string]*TestColumn
```

Comment:
how about the name of PGCatalogColumns instead?

Reply:
Done